crypto/chacha/asm/chacha-armv8-sve.pl

```diff
@@ -0,0 +1,1008 @@
+	&transpose($xa0,$xa1,$xa2,$xa3);
+	&transpose($xb0,$xb1,$xb2,$xb3);
+$code.=<<___;
+	eor	$xa0.d,$xa0.d,$xt0.d
+	eor	$xa1.d,$xa1.d,$xt1.d
+	eor	$xa2.d,$xa2.d,$xt2.d
+	eor	$xa3.d,$xa3.d,$xt3.d
+	eor	$xb0.d,$xb0.d,$xt4.d
+	eor	$xb1.d,$xb1.d,$xt5.d
+	eor	$xb2.d,$xb2.d,$xt6.d
+	eor	$xb3.d,$xb3.d,$xt7.d
+___
+	&transpose($xc0,$xc1,$xc2,$xc3);
+	&store_8x($xa0,$xa1,$xa2,$xa3,$xb0,$xb1,$xb2,$xb3);
+	&load_8x($xt0,$xt1,$xt2,$xt3,$xt4,$xt5,$xt6,$xt7);
+	&transpose($xd0,$xd1,$xd2,$xd3);
+$code.=<<___;
+	eor	$xc0.d,$xc0.d,$xt0.d
+	eor	$xc1.d,$xc1.d,$xt1.d
+	eor	$xc2.d,$xc2.d,$xt2.d
+	eor	$xc3.d,$xc3.d,$xt3.d
+	eor	$xd0.d,$xd0.d,$xt4.d
+	eor	$xd1.d,$xd1.d,$xt5.d
+	eor	$xd2.d,$xd2.d,$xt6.d
+	eor	$xd3.d,$xd3.d,$xt7.d
+___
+	&store_8x($xc0,$xc1,$xc2,$xc3,$xd0,$xd1,$xd2,$xd3);
+$code.=<<___;
+	add	$ctr,$ctr,$veclen
+100:
+___
+}
+}}}
+
+sub SVE_LOAD_STATES() {
+	my ($tmpw0,$tmpw1,$tmpw2,$tmpw3) = ("w9","w10","w11","w12");
+
+$code.=<<___;
+	index	@mx[12].s,$ictr,1
+	// FIXME following code are not functionally necessary
+	// but appear to be vital to perf test
+#if 1
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+	ptrues	p2.s,ALL
+#endif
+___
+$code.=<<___;
+	ldp	$tmpw0,$tmpw1,[$state]
+	ldp	$tmpw2,$tmpw3,[$state,#8]
+	dup	@mx[0].s,$tmpw0
+	dup	@mx[1].s,$tmpw1
+	dup	@mx[2].s,$tmpw2
+	dup	@mx[3].s,$tmpw3
+	ldp	$tmpw0,$tmpw1,[$state,#16]
+	ldp	$tmpw2,$tmpw3,[$state,#24]
+	dup	@mx[4].s,$tmpw0
+	dup	@mx[5].s,$tmpw1
+	dup	@mx[6].s,$tmpw2
+	dup	@mx[7].s,$tmpw3
+	ldp	$tmpw0,$tmpw1,[$state,#32]
+	ldp	$tmpw2,$tmpw3,[$state,#40]
+	dup	@mx[8].s,$tmpw0
+	dup	@mx[9].s,$tmpw1
+	dup	@mx[10].s,$tmpw2
+	dup	@mx[11].s,$tmpw3
+	ldp	$tmpw0,$tmpw1,[$state, #48]
+	ldp	$tmpw2,$tmpw3,[$state,#56]
+	dup	@mx[13].s,$tmpw1
+	dup	@mx[14].s,$tmpw2
+	dup	@mx[15].s,$tmpw3
+___
+}
+
+sub sve_handle_blocks() {
+	my ($counter) = ("x10");
+
+	&SVE_LOAD_STATES();
+
+$code.=<<___;
+	cmp	$blocks,$veclen
+	b.ge	2f
+	// predicated processing
+___
+$code.=<<___;
+	mov	$counter,#10
+.align	5
+1:
+___
+	&SVE_INNER_BLOCK(0);
+$code.=<<___;
+	subs	$counter,$counter,1
+	b.ne	1b
+	b	1f
+2:
+	// unpredicated processing
+___
+$code.=<<___;
+	mov	$counter,#10
+.align	5
+2:
+___
+	&SVE_INNER_BLOCK(1);
+$code.=<<___;
+	subs	$counter,$counter,1
+	b.ne	2b
+1:
+___
+	&SVE_TRANSFORMS();
+}
+
+sub chacha20_process() {
+	my ($counter) = ("x10");
+	my ($tmpw) = ("w11");
+
+$code.=<<___;
+	and	$len,$len,#63
+.align	5
+.Loop:
+	cbz	$blocks,.Ltail
+___
+	&sve_handle_blocks();
+$code.=<<___;
+	subs	$blocks,$blocks,$veclen
+	b.gt	.Loop
+.Ltail:
+	cbz	$len,.Lexit
+	cmp	$len,64
+	b.ge	2f
+	sub	sp,sp, 64
+	mov	$counter, 0
+1:
+	ldrb	$tmpw,[$inp,$counter]
+	strb	$tmpw,[sp,$counter]
+	add	$counter,$counter,1
+	cmp	$counter,$len
+	b.ne	1b
+	mov	$blocks,1
+	add	$len,$len,64
+	mov	$saved_outp,$outp
+	mov	$outp,sp
+	mov	$inp,sp
+	b	.Loop
+2:
+	sub	$len,$len,64
+2:
+	subs	$len,$len,1
+	ldrb	$tmpw, [sp, $len]
+	strb	$tmpw, [$saved_outp, $len]
+	b.gt	2b
+	add	sp, sp, 64
```

Comment:
So here it looks like if input is not a multiple of block_size you are copying the final partial input block to the stack, en/decrypting it in place, and then copying it back to the output.

I do not think this is necessary. If instead you rounded $blocks up to the nearest whole block (i.e. if $len&63 is non-zero, increment $blocks), and then predicated the load/stores in `output_regs` byte-wise ($dlen would have to be calculated from $len, rather than $blocks), then partial blocks would be handled in the main loop.
This should also be substantially faster for unusual-sized inputs.

If I'm missing something which makes the byte-wise predication in the main loop infeasible, then at least be sure to wipe the data copied to the stack when you copy out from it (otherwise we're potentially leaving sensitive data on the stack).

Reply:
i'll let the Neon handle the tail. SVE will no longer process tail.